### PR TITLE
cache metric data in metrics collector

### PR DESCRIPTION
**Pull request by boyang9527**

### CODE REVIEW

File: src/autoscaler/collection/TSDCache_test.go [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+					cache.Put(TestTSD{30, nil})
+
+					result, ok := cache.Query(30, 50, labels)
+					Expect(ok).To(BeTrue())
+					Expect(result).To(Equal([]TSD{TestTSD{30, nil}, TestTSD{40, nil}}))
+
+					cache.Put(TestTSD{50, nil})
+					result, ok = cache.Query(35, 50, labels)
+					Expect(ok).To(BeTrue())
+					Expect(result).To(Equal([]TSD{TestTSD{40, nil}}))
```
Comment (cdlliuy) [inline src/autoscaler/collection/TSDCache_test.go]: why query database directly, the "end" timestamp is included.  Is that possible to make the implementation compatible?

File: src/autoscaler/collection/TSDCache.go [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	} else {
+		l = c.cursor
+		r = c.cursor - 1 + c.capacity
+	}
+
+	for {
+		if l > r {
+			return l
+		}
+		m := (l + r) / 2
```
[the remark this answers is not in the document]
Comment (boyang9527) [reply src/autoscaler/collection/TSDCache.go]: fixed.

File: src/autoscaler/collection/TSDCache_test.go (hunk 1 above)
Comment (boyang9527) [reply src/autoscaler/collection/TSDCache_test.go]: Change made in https://github.com/cloudfoundry-incubator/app-autoscaler/pull/379/files#diff-689f77dbaf45e61a44746cb0b0ff27baR138 
mtrcs, ok := h.queryFunc(appId, start, end+1, order, labels)

use start, end+1 to query cache will make sure you get metrics data in [start, end]

**Comment by cdlliuy** [comment]
LGTM

**Pull request merged**